File: BaekjoonAlgorithm/Ruby/1463.rb
def check n
  array = Array.new(n+1, 0)
  (2..n).each do |number|
    array[number] = array[number-1] + 1
    if number % 3 == 0
      array[number] = [array[number], array[number/3]+1].min
    elsif number % 2 == 0
      array[number] = [array[number], array[number/2]+1].min
    end
  end
  array[n]
end
n = gets.to_i
print check(n) 

File: BaekjoonAlgorithm/Python/1463.py
d = [0] * 1000001
d[0] = 0
d[1] = 0
d[2] = 1

n = int(input())

for i in range(3,n+1) : 
  if(i % 6 == 0) :
    d[i] = min(min(d[i//2] + 1, d[i//3] + 1), d[i-1] + 1)
  elif(i % 3 == 0) :
    d[i] = min(d[i//3] + 1, d[i-1] + 1)
  elif(i % 2 == 0) :
    d[i] = min(d[i//2] + 1, d[i-1] + 1)
  else : 
    d[i] = d[i-1] + 1

print(d[n])
